pkg/kubeapiserver/admission/config.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"io/ioutil"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/config"
+	webhookconfig "k8s.io/apiserver/pkg/admission/plugin/webhook/config"
+	webhookinit "k8s.io/apiserver/pkg/admission/plugin/webhook/initializer"
+	"k8s.io/apiserver/pkg/server"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	cacheddiscovery "k8s.io/client-go/discovery/cached"
+	externalinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	aggregatorapiserver "k8s.io/kube-aggregator/pkg/apiserver"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	internalinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	quotainstall "k8s.io/kubernetes/pkg/quota/install"
+)
+
+type AdmissionConfig struct {
+	CloudConfigFile string
+
+	LoopbackClientConfig *rest.Config
+
+	ServiceResolver            aggregatorapiserver.ServiceResolver
+	WebhookAuthResolverWrapper config.AuthenticationInfoResolverWrapper
+
+	ExternalInformers externalinformers.SharedInformerFactory
+	InternalInformers internalinformers.SharedInformerFactory
+}
+
+func (c *AdmissionConfig) Complete(proxyTransport *http.Transport, enabledAggregatorRouting bool) completedAdmissionConfig {
```

Comment:
`Complete` usually usually only sets fields which are nil. Here though fields are unconditionally overridden. This looks more like a constructor than completion.

Reply:
yeah.. added checking nil before setting fields

Reply:
umm.. different from the fields inside `AdmissionConfig`, `proxyTransport` and `enabledAggregatorRouting` are sth/information extracted from aggregation layer. it's not an inside part of admission logic but more like an outside wrapper for proxy usage. an `AdmissionConfig` shouldn't contain extra information about aggregator, amiright?

Reply:
or what about move it as `New(proxyTransport, enabledAggregatorRouting)`?

Reply:
`New(proxyTransport, enabledAggregatorRouting)` is what I meant, yes.